src/UserSpaceInstrumentation/TrampolineTest.cpp

```diff
@@ -257,4 +264,213 @@ TEST(TrampolineTest, AllocateMemoryForTrampolines) {
   waitpid(pid, nullptr, 0);
 }
 
+TEST(TrampolineTest, AddressDifferenceAsInt32) {
+  // Result of the difference is negative; in the first case it just fits the second case overflows.
+  const uint64_t kAddr1 = 0x6012345612345678;
+  const uint64_t kAddr2Larger = kAddr1 - std::numeric_limits<int32_t>::min();
+  auto result = AddressDifferenceAsInt32(kAddr1, kAddr2Larger);
+  ASSERT_THAT(result, HasNoError());
+  EXPECT_EQ(std::numeric_limits<int32_t>::min(), result.value());
+  result = AddressDifferenceAsInt32(kAddr1, kAddr2Larger + 1);
+  EXPECT_THAT(result, HasError("Difference is larger than +-2GB"));
+
+  // Result of the difference is positive; in the first case it just fits the second case overflows.
+  const uint64_t kAddr2Smaller = kAddr1 - std::numeric_limits<int32_t>::max();
+  result = AddressDifferenceAsInt32(kAddr1, kAddr2Smaller);
+  ASSERT_THAT(result, HasNoError());
+  EXPECT_EQ(std::numeric_limits<int32_t>::max(), result.value());
+  result = AddressDifferenceAsInt32(kAddr1, kAddr2Smaller - 1);
+  EXPECT_THAT(result, HasError("Difference is larger than +-2GB"));
+
+  // Result of the difference does not even fit into a int64. We handle that gracefully as well.
+  const uint64_t kAddrHigh = 0xf234567812345678;
+  const uint64_t kAddrLow = kAddrHigh - 0xe234567812345678;
+  result = AddressDifferenceAsInt32(kAddrHigh, kAddrLow);
+  EXPECT_THAT(result, HasError("Difference is larger than +-2GB"));
+  result = AddressDifferenceAsInt32(kAddrLow, kAddrHigh);
+  EXPECT_THAT(result, HasError("Difference is larger than +-2GB"));
+}
+
+class RelocateInstructionTest : public testing::Test {
+ protected:
+  void SetUp() override {
+    CHECK(cs_open(CS_ARCH_X86, CS_MODE_64, &capstone_handle_) == CS_ERR_OK);
+    CHECK(cs_option(capstone_handle_, CS_OPT_DETAIL, CS_OPT_ON) == CS_ERR_OK);
+    instruction_ = cs_malloc(capstone_handle_);
+    CHECK(instruction_ != nullptr);
+  }
+
+  void Disassemble(const MachineCode& code) {
+    const uint8_t* code_pointer = code.GetResultAsVector().data();
+    size_t code_size = code.GetResultAsVector().size();
+    uint64_t disassemble_address = 0;
+    CHECK(cs_disasm_iter(capstone_handle_, &code_pointer, &code_size, &disassemble_address,
+                         instruction_));
+  }
+
+  void TearDown() override {
+    cs_free(instruction_, 1);
+    cs_close(&capstone_handle_);
+  }
+
+  cs_insn* instruction_ = nullptr;
+
+ private:
+  csh capstone_handle_ = 0;
+};
+
+TEST_F(RelocateInstructionTest, TrivialTranslation) {
+  MachineCode code;
+  // nop
+  code.AppendBytes({0x90});
+  Disassemble(code);
+
+  ErrorMessageOr<RelocatedInstruction> result =
+      RelocateInstruction(instruction_, 0x0100000000, 0x0200000000);
+  ASSERT_THAT(result, HasValue());
+  EXPECT_THAT(result.value().code, ElementsAreArray({0x90}));
+  EXPECT_FALSE(result.value().position_of_absolute_address.has_value());
+}
+
+TEST_F(RelocateInstructionTest, RipRelativeAddressing) {
```

Comment:
For all these tests, could you make it (even) more immediate to match what branch they refer to. I would suggest just a comment with the content of the "if".

Reply:
Added comments and moved trivial case to the bottom so they occur in the same order as in 'RelocateInstruction'